compiletest should be built with the stage2 compiler not the stage0 compiler

See #57709. Unfortunately the "fix" to that bug breaks the build when building using 1.31.0 and you have to unapply the patch to make it work against 1.31.0.

e.g. from https://buildd.debian.org/status/fetch.php?pkg=rustc&arch=amd64&ver=1.32.0%2Bdfsg1-1&stamp=1548672224&raw=0 :

~~~~
rustc command: "LD_LIBRARY_PATH"="/usr/lib:/<<BUILDDIR>>/rustc-1.32.0+dfsg1/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/release/deps:/usr/lib" "/usr/bin/rustc" "--crate-name" "compiletest" "
~~~~

The problem here is that if stage0 is 1.31.0 then the patch MUST be unapplied, if stage0 is 1.32.0 then the patch MUST be applied. Obviously we can't do both at once.

However in both Debian and Fedora we occasionally use either the current or the previous version of rustc to build a given version of rustc, and both of these versions should work.

cc @cuviper

Same issue with 1.40.0 again, when using itself as stage0:

~~~~
   Compiling compiletest v0.0.0 (/<<PKGBUILDDIR>>/src/tools/compiletest)
     Running `CARGO=/usr/bin/cargo CARGO_MANIFEST_DIR=/<<PKGBUILDDIR>>/src/tools/compiletest CARGO_PKG_VERSION=0.0.0 CARGO_PKG_VERSION_PATCH=0 CARGO_PKG_VERSION_MINOR=0 CARGO_PKG_VERSION_MAJOR=0 CARGO_PKG_NAME=compiletest CARGO_PKG_DESCRIPTION= CARGO_PKG_REPOSITORY= CARGO_PKG_AUTHORS='The Rust Project Developers' CARGO_PKG_HOMEPAGE= CARGO_PKG_VERSION_PRE= LD_LIBRARY_PATH='/<<PKGBUILDDIR>>/build/powerpc64le-unknown-linux-gnu/stage0-bootstrap-tools/release/deps:/usr/lib' /<<PKGBUILDDIR>>/build/bootstrap/debug/rustc --edition=2018 --crate-name compiletest src/tools/compiletest/src/main.rs --error-format json --json=diagnostic-rendered-ansi --crate-type bin --emit=dep-info,link -C opt-level=2 -C debuginfo=0 -C metadata=3bff92eb77ed034a -C extra-filename=-3bff92eb77ed034a --out-dir /<<PKGBUILDDIR>>/build/powerpc64le-unknown-linux-gnu/stage0-bootstrap-tools/powerpc64le-unknown-linux-gnu/release/deps --target powerpc64le-unknown-linux-gnu -C linker=powerpc64le-linux-gnu-gcc -L dependency=/<<PKGBUILDDIR>>/build/powerpc64le-unknown-linux-gnu/stage0-bootstrap-tools/powerpc64le-unknown-linux-gnu/release/deps -L dependency=/<<PKGBUILDDIR>>/build/powerpc64le-unknown-linux-gnu/stage0-bootstrap-tools/release/deps --extern diff=/<<PKGBUILDDIR>>/build/powerpc64le-unknown-linux-gnu/stage0-bootstrap-tools/powerpc64le-unknown-linux-gnu/release/deps/libdiff-f7c70ead82d3bb7b.rlib --extern env_logger=/<<PKGBUILDDIR>>/build/powerpc64le-unknown-linux-gnu/stage0-bootstrap-tools/powerpc64le-unknown-linux-gnu/release/deps/libenv_logger-42bd3e818814c443.rlib --extern getopts=/<<PKGBUILDDIR>>/build/powerpc64le-unknown-linux-gnu/stage0-bootstrap-tools/powerpc64le-unknown-linux-gnu/release/deps/libgetopts-9ad175f6b0afc4b5.rlib --extern lazy_static=/<<PKGBUILDDIR>>/build/powerpc64le-unknown-linux-gnu/stage0-bootstrap-tools/powerpc64le-unknown-linux-gnu/release/deps/liblazy_static-0d6c5f722411bcab.rlib --extern libc=/<<PKGBUILDDIR>>/build/powerpc64le-unknown-linux-gnu/stage0-bootstrap-tools/powerpc64le-unknown-linux-gnu/release/deps/liblibc-0f34c56984de760b.rlib --extern log=/<<PKGBUILDDIR>>/build/powerpc64le-unknown-linux-gnu/stage0-bootstrap-tools/powerpc64le-unknown-linux-gnu/release/deps/liblog-1d69c45aedc3973c.rlib --extern regex=/<<PKGBUILDDIR>>/build/powerpc64le-unknown-linux-gnu/stage0-bootstrap-tools/powerpc64le-unknown-linux-gnu/release/deps/libregex-877563e5fb3e2b70.rlib --extern rustfix=/<<PKGBUILDDIR>>/build/powerpc64le-unknown-linux-gnu/stage0-bootstrap-tools/powerpc64le-unknown-linux-gnu/release/deps/librustfix-52c76b39e6b4bcbd.rlib --extern serde=/<<PKGBUILDDIR>>/build/powerpc64le-unknown-linux-gnu/stage0-bootstrap-tools/powerpc64le-unknown-linux-gnu/release/deps/libserde-6b72e6c8366be6dd.rlib --extern serde_json=/<<PKGBUILDDIR>>/build/powerpc64le-unknown-linux-gnu/stage0-bootstrap-tools/powerpc64le-unknown-linux-gnu/release/deps/libserde_json-0db31fa1a064a5dd.rlib --extern walkdir=/<<PKGBUILDDIR>>/build/powerpc64le-unknown-linux-gnu/stage0-bootstrap-tools/powerpc64le-unknown-linux-gnu/release/deps/libwalkdir-243dfb4a0f7f76f8.rlib -C link-args=-Wl,-z,relro --cap-lints warn --remap-path-prefix=/<<PKGBUILDDIR>>=/usr/src/rustc-1.40.0 -Zexternal-macro-backtrace -Zbinary-dep-depinfo -L native=/<<PKGBUILDDIR>>/build/powerpc64le-unknown-linux-gnu/stage0-bootstrap-tools/powerpc64le-unknown-linux-gnu/release/build/backtrace-sys-6a5e41a89dbb2581/out`
rustc command: "LD_LIBRARY_PATH"="/usr/lib:/<<PKGBUILDDIR>>/build/powerpc64le-unknown-linux-gnu/stage0-bootstrap-tools/release/deps:/usr/lib" "/usr/bin/rustc" "--edition=2018" "--crate-name" "compiletest" "src/tools/compiletest/src/main.rs" "--error-format" "json" "--json=diagnostic-rendered-ansi" "--crate-type" "bin" "--emit=dep-info,link" "-C" "opt-level=2" "-C" "debuginfo=0" "-C" "metadata=3bff92eb77ed034a" "-C" "extra-filename=-3bff92eb77ed034a" "--out-dir" "/<<PKGBUILDDIR>>/build/powerpc64le-unknown-linux-gnu/stage0-bootstrap-tools/powerpc64le-unknown-linux-gnu/release/deps" "--target" "powerpc64le-unknown-linux-gnu" "-C" "linker=powerpc64le-linux-gnu-gcc" "-L" "dependency=/<<PKGBUILDDIR>>/build/powerpc64le-unknown-linux-gnu/stage0-bootstrap-tools/powerpc64le-unknown-linux-gnu/release/deps" "-L" "dependency=/<<PKGBUILDDIR>>/build/powerpc64le-unknown-linux-gnu/stage0-bootstrap-tools/release/deps" "--extern" "diff=/<<PKGBUILDDIR>>/build/powerpc64le-unknown-linux-gnu/stage0-bootstrap-tools/powerpc64le-unknown-linux-gnu/release/deps/libdiff-f7c70ead82d3bb7b.rlib" "--extern" "env_logger=/<<PKGBUILDDIR>>/build/powerpc64le-unknown-linux-gnu/stage0-bootstrap-tools/powerpc64le-unknown-linux-gnu/release/deps/libenv_logger-42bd3e818814c443.rlib" "--extern" "getopts=/<<PKGBUILDDIR>>/build/powerpc64le-unknown-linux-gnu/stage0-bootstrap-tools/powerpc64le-unknown-linux-gnu/release/deps/libgetopts-9ad175f6b0afc4b5.rlib" "--extern" "lazy_static=/<<PKGBUILDDIR>>/build/powerpc64le-unknown-linux-gnu/stage0-bootstrap-tools/powerpc64le-unknown-linux-gnu/release/deps/liblazy_static-0d6c5f722411bcab.rlib" "--extern" "libc=/<<PKGBUILDDIR>>/build/powerpc64le-unknown-linux-gnu/stage0-bootstrap-tools/powerpc64le-unknown-linux-gnu/release/deps/liblibc-0f34c56984de760b.rlib" "--extern" "log=/<<PKGBUILDDIR>>/build/powerpc64le-unknown-linux-gnu/stage0-bootstrap-tools/powerpc64le-unknown-linux-gnu/release/deps/liblog-1d69c45aedc3973c.rlib" "--extern" "regex=/<<PKGBUILDDIR>>/build/powerpc64le-unknown-linux-gnu/stage0-bootstrap-tools/powerpc64le-unknown-linux-gnu/release/deps/libregex-877563e5fb3e2b70.rlib" "--extern" "rustfix=/<<PKGBUILDDIR>>/build/powerpc64le-unknown-linux-gnu/stage0-bootstrap-tools/powerpc64le-unknown-linux-gnu/release/deps/librustfix-52c76b39e6b4bcbd.rlib" "--extern" "serde=/<<PKGBUILDDIR>>/build/powerpc64le-unknown-linux-gnu/stage0-bootstrap-tools/powerpc64le-unknown-linux-gnu/release/deps/libserde-6b72e6c8366be6dd.rlib" "--extern" "serde_json=/<<PKGBUILDDIR>>/build/powerpc64le-unknown-linux-gnu/stage0-bootstrap-tools/powerpc64le-unknown-linux-gnu/release/deps/libserde_json-0db31fa1a064a5dd.rlib" "--extern" "walkdir=/<<PKGBUILDDIR>>/build/powerpc64le-unknown-linux-gnu/stage0-bootstrap-tools/powerpc64le-unknown-linux-gnu/release/deps/libwalkdir-243dfb4a0f7f76f8.rlib" "-C" "link-args=-Wl,-z,relro" "--cap-lints" "warn" "--remap-path-prefix=/<<PKGBUILDDIR>>=/usr/src/rustc-1.40.0" "-Zexternal-macro-backtrace" "-Zbinary-dep-depinfo" "-L" "native=/<<PKGBUILDDIR>>/build/powerpc64le-unknown-linux-gnu/stage0-bootstrap-tools/powerpc64le-unknown-linux-gnu/release/build/backtrace-sys-6a5e41a89dbb2581/out" "--sysroot" "/usr" "-C" "debug-assertions=n"
sysroot: "/usr"
libdir: "/usr/lib"
error[E0063]: missing field `time_options` in initializer of `test::TestOpts`
   --> src/tools/compiletest/src/main.rs:545:5
    |
545 |     test::TestOpts {
    |     ^^^^^^^^^^^^^^ missing `time_options`

error[E0063]: missing field `test_type` in initializer of `test::TestDesc`
   --> src/tools/compiletest/src/main.rs:701:23
    |
701 |                 desc: test::TestDesc {
    |                       ^^^^^^^^^^^^^^ missing `test_type`

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0063`.

Did not run successfully: Ok(ExitStatus(ExitStatus(256)))
"/usr/bin/rustc" "--edition=2018" "--crate-name" "compiletest" "src/tools/compiletest/src/main.rs" "--error-format" "json" "--json=diagnostic-rendered-ansi" "--crate-type" "bin" "--emit=dep-info,link" "-C" "opt-level=2" "-C" "debuginfo=0" "-C" "metadata=3bff92eb77ed034a" "-C" "extra-filename=-3bff92eb77ed034a" "--out-dir" "/<<PKGBUILDDIR>>/build/powerpc64le-unknown-linux-gnu/stage0-bootstrap-tools/powerpc64le-unknown-linux-gnu/release/deps" "--target" "powerpc64le-unknown-linux-gnu" "-C" "linker=powerpc64le-linux-gnu-gcc" "-L" "dependency=/<<PKGBUILDDIR>>/build/powerpc64le-unknown-linux-gnu/stage0-bootstrap-tools/powerpc64le-unknown-linux-gnu/release/deps" "-L" "dependency=/<<PKGBUILDDIR>>/build/powerpc64le-unknown-linux-gnu/stage0-bootstrap-tools/release/deps" "--extern" "diff=/<<PKGBUILDDIR>>/build/powerpc64le-unknown-linux-gnu/stage0-bootstrap-tools/powerpc64le-unknown-linux-gnu/release/deps/libdiff-f7c70ead82d3bb7b.rlib" "--extern" "env_logger=/<<PKGBUILDDIR>>/build/powerpc64le-unknown-linux-gnu/stage0-bootstrap-tools/powerpc64le-unknown-linux-gnu/release/deps/libenv_logger-42bd3e818814c443.rlib" "--extern" "getopts=/<<PKGBUILDDIR>>/build/powerpc64le-unknown-linux-gnu/stage0-bootstrap-tools/powerpc64le-unknown-linux-gnu/release/deps/libgetopts-9ad175f6b0afc4b5.rlib" "--extern" "lazy_static=/<<PKGBUILDDIR>>/build/powerpc64le-unknown-linux-gnu/stage0-bootstrap-tools/powerpc64le-unknown-linux-gnu/release/deps/liblazy_static-0d6c5f722411bcab.rlib" "--extern" "libc=/<<PKGBUILDDIR>>/build/powerpc64le-unknown-linux-gnu/stage0-bootstrap-tools/powerpc64le-unknown-linux-gnu/release/deps/liblibc-0f34c56984de760b.rlib" "--extern" "log=/<<PKGBUILDDIR>>/build/powerpc64le-unknown-linux-gnu/stage0-bootstrap-tools/powerpc64le-unknown-linux-gnu/release/deps/liblog-1d69c45aedc3973c.rlib" "--extern" "regex=/<<PKGBUILDDIR>>/build/powerpc64le-unknown-linux-gnu/stage0-bootstrap-tools/powerpc64le-unknown-linux-gnu/release/deps/libregex-877563e5fb3e2b70.rlib" "--extern" "rustfix=/<<PKGBUILDDIR>>/build/powerpc64le-unknown-linux-gnu/stage0-bootstrap-tools/powerpc64le-unknown-linux-gnu/release/deps/librustfix-52c76b39e6b4bcbd.rlib" "--extern" "serde=/<<PKGBUILDDIR>>/build/powerpc64le-unknown-linux-gnu/stage0-bootstrap-tools/powerpc64le-unknown-linux-gnu/release/deps/libserde-6b72e6c8366be6dd.rlib" "--extern" "serde_json=/<<PKGBUILDDIR>>/build/powerpc64le-unknown-linux-gnu/stage0-bootstrap-tools/powerpc64le-unknown-linux-gnu/release/deps/libserde_json-0db31fa1a064a5dd.rlib" "--extern" "walkdir=/<<PKGBUILDDIR>>/build/powerpc64le-unknown-linux-gnu/stage0-bootstrap-tools/powerpc64le-unknown-linux-gnu/release/deps/libwalkdir-243dfb4a0f7f76f8.rlib" "-C" "link-args=-Wl,-z,relro" "--cap-lints" "warn" "--remap-path-prefix=/<<PKGBUILDDIR>>=/usr/src/rustc-1.40.0" "-Zexternal-macro-backtrace" "-Zbinary-dep-depinfo" "-L" "native=/<<PKGBUILDDIR>>/build/powerpc64le-unknown-linux-gnu/stage0-bootstrap-tools/powerpc64le-unknown-linux-gnu/release/build/backtrace-sys-6a5e41a89dbb2581/out" "--sysroot" "/usr" "-C" "debug-assertions=n"
~~~~

CC @cuviper AFAICT you do not have a patch that fixes this on Fedora either.

--- a/src/bootstrap/tool.rs
+++ b/src/bootstrap/tool.rs
@@ -290,6 +290,7 @@ macro_rules! bootstrap_tool {
     ($(
         $name:ident, $path:expr, $tool_name:expr
         $(,is_external_tool = $external:expr)*
+        $(,is_unstable_tool = $unstable:expr)*
         $(,features = $features:expr)*
         ;
     )+) => {
@@ -340,7 +341,12 @@ macro_rules! bootstrap_tool {
                     compiler: self.compiler,
                     target: self.target,
                     tool: $tool_name,
-                    mode: Mode::ToolBootstrap,
+                    mode: if false $(|| $unstable)* {
+                        // use in-tree libraries for unstable features
+                        Mode::ToolStd
+                    } else {
+                        Mode::ToolBootstrap
+                    },
                     path: $path,
                     is_optional_tool: false,
                     source_type: if false $(|| $external)* {
@@ -367,7 +373,7 @@ bootstrap_tool!(
     Tidy, "src/tools/tidy", "tidy";
     Linkchecker, "src/tools/linkchecker", "linkchecker";
     CargoTest, "src/tools/cargotest", "cargotest";
-    Compiletest, "src/tools/compiletest", "compiletest";
+    Compiletest, "src/tools/compiletest", "compiletest", is_unstable_tool = true;
     BuildManifest, "src/tools/build-manifest", "build-manifest";
     RemoteTestClient, "src/tools/remote-test-client", "remote-test-client";
     RustInstaller, "src/tools/rust-installer", "fabricate", is_external_tool = true;

--- a/src/bootstrap/tool.rs
+++ b/src/bootstrap/tool.rs
@@ -289,7 +289,6 @@ fn rustbook_features() -> Vec<String> {
 macro_rules! bootstrap_tool {
     ($(
         $name:ident, $path:expr, $tool_name:expr
-        $(,llvm_tools = $llvm:expr)*
         $(,is_external_tool = $external:expr)*
         $(,features = $features:expr)*
         ;
@@ -299,15 +298,6 @@ macro_rules! bootstrap_tool {
             $(
                 $name,
             )+
-        }
-
-        impl Tool {
-            /// Whether this tool requires LLVM to run
-            pub fn uses_llvm_tools(&self) -> bool {
-                match self {
-                    $(Tool::$name => false $(|| $llvm)*,)+
-                }
-            }
         }
 
         impl<'a> Builder<'a> {
@@ -377,7 +367,7 @@ bootstrap_tool!(
     Tidy, "src/tools/tidy", "tidy";
     Linkchecker, "src/tools/linkchecker", "linkchecker";
     CargoTest, "src/tools/cargotest", "cargotest";
-    Compiletest, "src/tools/compiletest", "compiletest", llvm_tools = true;
+    Compiletest, "src/tools/compiletest", "compiletest";
     BuildManifest, "src/tools/build-manifest", "build-manifest";
     RemoteTestClient, "src/tools/remote-test-client", "remote-test-client";
     RustInstaller, "src/tools/rust-installer", "fabricate", is_external_tool = true;

--- a/src/tools/compiletest/src/main.rs
+++ b/src/tools/compiletest/src/main.rs
@@ -1,6 +1,8 @@
 #![crate_name = "compiletest"]
-#![feature(test)]
 #![deny(warnings)]
+// The `test` crate is the only unstable feature
+// allowed here, just to share similar code.
+#![feature(test)]
 
 extern crate test;
 

